# Intermediate Assessment
## By Gladwin Thomas
## DS09

## Problem Description:
A Chinese automobile company aspires to enter the US market by setting up their manufacturing unit
there and producing cars locally to give competition to their US and European counterparts. They have
contracted an automobile consulting company to understand the factors on which the pricing of cars
depends. Specifically, they want to understand the factors affecting the pricing of cars in the American
market, since those may be very different from the Chinese market. 

### Essentially, the company wants to
know:
Which variables are significant in predicting the price of a car?
How well those variables describe the price of a car?
Based on various market surveys, the consulting firm has gathered a large dataset of different types of
cars across the American market.

### Business Goal: 
You are required to model the price of cars with the available independent variables. It
will be used by the management to understand how exactly the prices vary with the independent
variables. They can accordingly manipulate the design of the cars, the business strategy, etc., to meet
certain price levels. Further, the model will be a good way for the management to understand the pricing
dynamics of a new market.

## Dataset:


### Key components to be fulfilled:
1. Loading and Preprocessing (5 marks): Load the dataset and perform necessary preprocessing
steps.
2. Model Implementation (10 marks): Implement the following five regression algorithms:
● Linear Regression
● Decision Tree Regressor
● Random Forest Regressor
● Gradient Boosting Regressor
● Support Vector Regressor
3. Model Evaluation (5 marks): Compare the performance of all the models based on R-squared,
Mean Squared Error (MSE), and Mean Absolute Error (MAE). Identify the best performing model
and justify why it is the best.
4. Feature Importance Analysis (2 marks): Identify the significant variables affecting car prices
(feature selection)
5. Hyperparameter Tuning (2 marks): Perform hyperparameter tuning and check whether the
performance of the model has increased.
6. Timely Submission (1 mark) Submission Guidelines: - Provide your code in a Jupyter Notebook
format and submit the GitHub link here. - Ensure your explanations and answers are clear and
concise.

Total Score: 25

In [1]:
# Import Libraries
import pandas as pd
import numpy as np

In [2]:
# Read the dataset
df = pd.read_csv("CarPrice_Assignment.csv")

In [3]:
# Top 5 rowss
df.head()

,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0


In [4]:
# Shape of the dataset
df.shape

(205, 26)

In [5]:
# 50 Samples of data to check any data discrepancies
df.sample(50)

,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
38,39,0,honda civic 1300,gas,std,two,hatchback,fwd,front,96.5,...,110,1bbl,3.15,3.58,9.0,86,5800,27,33,9095.0
163,164,1,toyota corolla liftback,gas,std,two,sedan,rwd,front,94.5,...,98,2bbl,3.19,3.03,9.0,70,4800,29,34,8058.0
169,170,2,toyota starlet,gas,std,two,hatchback,rwd,front,98.4,...,146,mpfi,3.62,3.50,9.3,116,4800,24,30,9989.0
112,113,0,peugeot 604sl,diesel,turbo,four,sedan,rwd,front,107.9,...,152,idi,3.70,3.52,21.0,95,4150,28,33,16900.0
51,52,1,maxda glc deluxe,gas,std,two,hatchback,fwd,front,93.1,...,91,2bbl,3.03,3.15,9.0,68,5000,31,38,6095.0
69,70,0,buick century,diesel,turbo,two,hardtop,rwd,front,106.7,...,183,idi,3.58,3.64,21.5,123,4350,22,25,28176.0
180,181,-1,toyota starlet,gas,std,four,sedan,rwd,front,104.5,...,171,mpfi,3.27,3.35,9.2,156,5200,20,24,15690.0
134,135,3,saab 99le,gas,std,two,hatchback,fwd,front,99.1,...,121,mpfi,2.54,2.07,9.3,110,5250,21,28,15040.0
177,178,-1,toyota mark ii,gas,std,four,hatchback,fwd,front,102.4,...,122,mpfi,3.31,3.54,8.7,92,4200,27,32,11248.0
24,25,1,dodge monaco (sw),gas,std,four,hatchback,fwd,front,93.7,...,90,2bbl,2.97,3.23,9.4,68,5500,31,38,6229.0


In [6]:
### Observation - No discrepancies found in the data

In [7]:
# Check null values
df.isna().sum()

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

In [8]:
# Unique values present in each column
for col in df.columns:
    print(col)
    print(df[col].unique())

car_ID
[  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
 199 200 201 202 203 204 205]
symboling
[ 3  1  2  0 -1 -2]
CarName
['alfa-romero giulia' 'alfa-romero stelvio' 'alfa-romero Quadrifoglio'
 'audi 100 ls' 'audi 100ls' 'audi fox' 'audi 5000' 

#### Observation:  No null values found

In [9]:
# Import Libraries
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer

In [10]:
# Column names convert to lowercase
df.columns = df.columns.str.lower()

In [11]:
# Car Brands are extracted from carname
df['brand'] = df["carname"].apply(lambda x: x.split()[0])
df.brand.unique()

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'Nissan', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

In [12]:
# Dropping car_id and carname, as it doesn't help in price predictions
df.drop(['car_id', 'carname'], axis=1, inplace=True)

In [13]:
# Get the features or independent variables, X
X = df.drop('price',axis=1)
X.columns

Index(['symboling', 'fueltype', 'aspiration', 'doornumber', 'carbody',
       'drivewheel', 'enginelocation', 'wheelbase', 'carlength', 'carwidth',
       'carheight', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize',
       'fuelsystem', 'boreratio', 'stroke', 'compressionratio', 'horsepower',
       'peakrpm', 'citympg', 'highwaympg', 'brand'],
      dtype='object')

In [14]:
# Get the target variable or dependent variable, y
y = df['price']
y

0      13495.0
1      16500.0
2      16500.0
3      13950.0
4      17450.0
        ...   
200    16845.0
201    19045.0
202    21485.0
203    22470.0
204    22625.0
Name: price, Length: 205, dtype: float64

# Preprocessing Pipeline
Creating a pipeline for preprocessing for easier model training and resuse.

In [15]:
# Getting categorical and numerical columns
categorical_cols = X.select_dtypes(include="object").columns
numerical_cols = X.select_dtypes(exclude="object").columns

# Creating a preprocessor pipeline for columns
preprocessor = ColumnTransformer(
        # List of transformers
    transformers=[
        # For numerical columns (name, transformers, columns)
        (
            "num",
            Pipeline([
                ("imputer", SimpleImputer(strategy="median")),
                ("scaler", StandardScaler())
            ]),
            numerical_cols
        ),
        # For categorical columns (name, transformers, columns)
        (
            "cat",
            Pipeline([
                ("imputer", SimpleImputer(strategy="most_frequent")),
                ("encoder", OneHotEncoder(handle_unknown="ignore"))
            ]),
            categorical_cols
        )
    ]
)

## Train Test Split

In [16]:
# Train test split of X and y
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

# Model Implementation

## Linear Regression


In [17]:
from sklearn.linear_model import LinearRegression

# Linear Regression pipeline consisting data preprocessing and model training
lr_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", LinearRegression())
])

## Decision Tree Regressor

In [18]:
from sklearn.tree import DecisionTreeRegressor

# Decision Tree pipeline consisting data preprocessing and model training

dt_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", DecisionTreeRegressor(random_state=42))
])

## Random Forest

In [19]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest pipeline consisting data preprocessing and model training

rf_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", RandomForestRegressor(
        n_estimators=200,
        random_state=42
    ))
])

## Gradient Boosting Regressor

In [20]:
from sklearn.ensemble import GradientBoostingRegressor

# Gradient Boosting pipeline consisting data preprocessing and model training

gb_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", GradientBoostingRegressor(random_state=42))
])

## Support Vector Regressor

In [21]:
from sklearn.svm import SVR

# Support Vector Regressor pipeline consisting data preprocessing and model training

svr_model = Pipeline([
    ("preprocessor", preprocessor),
    ("model", SVR())
])

# Model Evaluation

In [22]:
# Performing model evaluation after model training using r2_score, mse and mae

from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Defining the model dictionary
models = {
    "Linear Regression": lr_model,
    "Decision Tree": dt_model,
    "Random Forest": rf_model,
    "Gradient Boosting": gb_model,
    "SVR": svr_model
}

results = []

# Loop for training and evaluating the models
for name, model in models.items():

    model.fit(X_train, y_train)

    pred = model.predict(X_test)

    results.append([
        name,
        r2_score(y_test, pred),
        mean_squared_error(y_test, pred),
        mean_absolute_error(y_test, pred)
    ])

# Results are made to a dataframe for easier access 
results_df = pd.DataFrame(
    results,
    columns=["Model", "R2", "MSE", "MAE"]
)

# Results are printed
print(results_df.sort_values("R2", ascending=False))

               Model        R2           MSE          MAE
2      Random Forest  0.958881  3.246093e+06  1275.282101
3  Gradient Boosting  0.926791  5.779426e+06  1695.764682
0  Linear Regression  0.899880  7.903858e+06  1835.332222
1      Decision Tree  0.896224  8.192494e+06  1923.678854
4                SVR -0.099789  8.682180e+07  5695.277764


## Observation

Random Forest Regressor is the best-performing model because:

- Highest R² score (0.958)
- Lowest MSE
- Lowest MAE


# Feature Importance Analysis

In [23]:
# Since random forest is the best model, we take feature importance from random forest

# Training the model
rf_model.fit(X_train, y_train)

# Getting the feature names from the random forest pipeline
feature_names = rf_model.named_steps[
    "preprocessor"
].get_feature_names_out()

# Getting the feature importances from the random forest pipeline
importance = rf_model.named_steps[
    "model"
].feature_importances_

# Creating a df using importance and feature names
importance_df = pd.DataFrame({
    "Feature": feature_names,
    "Importance": importance
})

# Sort values importance
importance_df.sort_values(
    by="Importance",
    ascending=False
).head(10)

,Feature,Importance
6,num__enginesize,0.584096
5,num__curbweight,0.268298
13,num__highwaympg,0.042013
10,num__horsepower,0.023676
3,num__carwidth,0.012716
2,num__carlength,0.008479
1,num__wheelbase,0.007321
54,cat__brand_bmw,0.007192
12,num__citympg,0.006139
11,num__peakrpm,0.005626


## Observation


### We can see that the importance of features are:
    1. engine size - 58.4%
    2. Curb weight - 26.8%
    3. Highway mpg - 4.2%
    4. Horsepower  - 2.3%
    5. Car width   - 1.2%

# Hyper Parameter Tuning

In [24]:
# Import GridSearchCV for hyperparameter tuning

from sklearn.model_selection import GridSearchCV

# Setting the parameters to be checked
param_grid = {
    "model__n_estimators": [100, 200, 300],
    "model__max_depth": [None, 10, 20],
    "model__min_samples_split": [2, 5]
}

# Initializing the gridsearchcv
grid = GridSearchCV(
    rf_model,
    param_grid,
    cv=5,
    scoring="r2",
    n_jobs=-1
)

# Training the gridsearchcv
grid.fit(X_train, y_train)

# Best parameter for the random forest model
print(grid.best_params_)

{'model__max_depth': 10, 'model__min_samples_split': 2, 'model__n_estimators': 300}


In [25]:
best_rf = grid.best_estimator_

pred = best_rf.predict(X_test)

print("Performance before tuning of Random Forest: \nR2 - 0.958881 MSE - 3.246093e+06 MAE - 1275.282101\n")
print("Performance after tuning of Random Forest: ")
print("R2:", r2_score(y_test, pred))
print("MSE", mean_squared_error(y_test, pred))
print("MAE:", mean_absolute_error(y_test, pred))

Performance before tuning of Random Forest: 
R2 - 0.958881 MSE - 3.246093e+06 MAE - 1275.282101

Performance after tuning of Random Forest: 
R2: 0.9590546780180843
MSE 3232388.353464655
MAE: 1230.72116662515
